Editorial review (#32)
fcalabri/bmwg-ai-fabric-terminology: draft-calabria-bmwg-ai-fabric-terminology.md

## Relationship to Existing BMWG Work

This document extends the foundational BMWG terminology established
in {{!RFC1242}} (network interconnect benchmarking terminology) and
{{!RFC8238}} (data center benchmarking terminology). Where terms are
defined in those RFCs, this document provides AI fabric context
extensions; the core definitions remain as established. This document

> also extends the test methodology framework of {{!RFC2544}} and

Reviewer: It looks extension to the test methodology framework should be defined in AI fabric methodology draft instead of this terminology draft, remove?

Author: wanted to add it to all three documents, started with this one - the training and inference should follow,  

would you recommend removing it from here and adding it ONLY to training and inference?  that'd work for me, please advise

Reviewer: Based on my experience, keeping it ONLY in the training and inference makes sense.

| **DUT** | Device Under Test. The network element(s) whose performance characteristics are being measured. In AI fabric benchmarking the DUT is one or more fabric elements: leaf switches, spine switches, NICs, or the complete fabric assembly. |
| **SUT** | System Under Test. The complete AI compute system including accelerators, NICs, the fabric DUT, and serving/training software, when end-to-end metrics are the measurement objective. |
| **RT** | Router Tester / Traffic Generator. Test equipment capable of generating and receiving network traffic at specified rates with nanosecond-resolution timestamping sufficient for the measurements defined in the companion methodology documents. |

> | **JFI** | Jain's Fairness Index. A scalar measure of flow-level throughput fairness across n flows: `JFI = (Σxᵢ)² / (n · Σxᵢ²)` where xᵢ is the throughput of flow i. A value of 1.0 indicates perfect fairness; lower values indicate disparity. Defined in {{Jain1984}}. **SHOULD** be reported alongside throughput measurements for all multi-flow AI fabric tests. |

Reviewer: ```suggestion
| **JFI** | Fairness Index. A scalar measure of flow-level throughput fairness across n flows {{Jain1984}}: `JFI = (Σxᵢ)² / (n · Σxᵢ²)` where xᵢ is the throughput of flow i. A value of 1.0 indicates perfect fairness; lower values indicate disparity. **SHOULD** be reported alongside throughput measurements for all multi-flow AI fabric tests. |
```

| **AllGather** | A collective in which each participant contributes a shard of a tensor and all participants receive the concatenation of all shards. Used in tensor-parallel (Megatron-style) layers to reconstruct distributed activations or parameters. |
| **ReduceScatter** | A collective combining an element-wise reduction with a scatter, so each participant receives a distinct slice of the reduced result. Used in ZeRO-stage optimizer strategies and as the first half of a ring-AllReduce. |
| **AllToAll** | A collective in which each participant sends a distinct payload to every other participant and receives a distinct payload from every other participant. The critical collective for Mixture-of-Experts token dispatch. Generates N(N−1) independent point-to-point flows for N participants. |
| **Ring Algorithm** | An AllReduce (or AllGather/ReduceScatter) algorithm structured as a logical ring of participants. Each participant sends to its right neighbor and receives from its left neighbor in 2(N−1) steps. Bus bandwidth efficiency = 2(N−1)/N, approaching 100% for large N. Standard baseline for BusBW calculation. |

> | **BusBW** | The effective data throughput per accelerator during a collective operation, computed as:<br/><br/>BusBW = (data_size × algo_factor) / time<br/><br/>algo_factor is a fixed normalization constant derived from the ideal ring algorithm for each collective type, applied regardless of the algorithm actually selected by the collective library at runtime. This makes BusBW algorithm-invariant: the same hardware moving the same data volume in the same time yields the same BusBW whether the library selects ring, tree, or recursive doubling. This convention follows the NCCL nccl-tests definition.<br/><br/>Collective       algo_factor<br/>AllReduce        2 × (n−1) / n<br/>AllGather        (n−1) / n<br/>ReduceScatter    (n−1) / n<br/>AllToAll         (n−1) / n<br/><br/>n = number of participating accelerators.<br/><br/>Worked example — AllReduce, n=8, data_size=1 GB, time=10 ms:<br/>algo_factor = 2 × (8−1) / 8 = 1.75<br/>BusBW = (1 GB × 1.75) / 10 ms = 175 GB/s<br/><br/>Reports MUST state: collective type, algo_factor value, collective library name and version, and n. The algorithm actually selected by the library SHOULD be reported as diagnostic information when known. Units: GB/s or Gbps; reports MUST state which. |

Reviewer: Do we want to make make BusBw calculation strongly dependson NVIDIA Collective Communications Library (NCCL)?

Author: wanted to use it as a guideline as we received feedback that BusBW algo_factor was  ambiguous ,  and makes sense since the whole point of BusBW is being algorithm-invariant. The original definition contradicted this and may  produce inconsistent reports.

See these examples from the previous definition-> 

-> The same hardware moving the same bytes in the same  time can report different BusBW just because NCCL switched from tree to ring at the size-threshold boundary.
-> A library upgrade (e.g., NCCL 2.18 → 2.22) re-tunes thresholds and shifts the BusBW curve without any hardware change.
-> A vendor that ships a custom CCL with a more aggressive tree algorithm appears "faster" in BusBW terms even when on-wire bytes per second are identical.

So on this new definition a   fixed-ring algo_factor convention considers these situations, , so  for example benchmarkiing two fabrics / DUTs  that move the same data in the same time provide same results.

> | **BusBW** | The effective data throughput per accelerator during a collective operation, computed as:<br/><br/>BusBW = (data_size × algo_factor) / time<br/><br/>algo_factor is a fixed normalization constant derived from the ideal ring algorithm for each collective type, applied regardless of the algorithm actually selected by the collective library at runtime. This makes BusBW algorithm-invariant: the same hardware moving the same data volume in the same time yields the same BusBW whether the library selects ring, tree, or recursive doubling. This convention follows the NCCL nccl-tests definition.<br/><br/>Collective       algo_factor<br/>AllReduce        2 × (n−1) / n<br/>AllGather        (n−1) / n<br/>ReduceScatter    (n−1) / n<br/>AllToAll         (n−1) / n<br/><br/>n = number of participating accelerators.<br/><br/>Worked example — AllReduce, n=8, data_size=1 GB, time=10 ms:<br/>algo_factor = 2 × (8−1) / 8 = 1.75<br/>BusBW = (1 GB × 1.75) / 10 ms = 175 GB/s<br/><br/>Reports MUST state: collective type, algo_factor value, collective library name and version, and n. The algorithm actually selected by the library SHOULD be reported as diagnostic information when known. Units: GB/s or Gbps; reports MUST state which. |

Reviewer: ```suggestion
| **BusBW** | The effective data throughput per accelerator during a collective operation, computed as:<br/><br/>BusBW = (data_size × algo_factor) / time<br/><br/>algo_factor is a fixed normalization constant derived from the ideal ring algorithm for each collective type, applied regardless of the algorithm actually selected by the collective library at runtime. This makes BusBW algorithm-invariant: the same hardware moving the same data volume in the same time yields the same BusBW whether the library selects ring, tree, or recursive doubling. The algo_factor calculation MUST conform to the formula specified here.<br/><br/>Collective       algo_factor<br/>AllReduce        2 × (n−1) / n<br/>AllGather        (n−1) / n<br/>ReduceScatter    (n−1) / n<br/>AllToAll         (n−1) / n<br/><br/>n = number of participating accelerators.<br/><br/>Worked example — AllReduce, n=8, data_size=1 GB, time=10 ms:<br/>algo_factor = 2 × (8−1) / 8 = 1.75<br/>BusBW = (1 GB × 1.75) / 10 ms = 175 GB/s<br/><br/>Reports MUST state: collective type, algo_factor value, collective library name and version, and n. The algorithm actually selected by the library SHOULD be reported as diagnostic information when known. Units: GB/s or Gbps; reports MUST state which. |
```

| **RFC 8238** | "Data Center Benchmarking Terminology" (Bitar et al., 2017). Extends RFC 1242 with data center-relevant terms including forwarding table scaling, congestion, and VM/SDN. Incast, ECN, and buffer occupancy concepts in this document align with RFC 8238 definitions. |
| **RFC 8239** | "Data Center Benchmarking Methodology" (Bitar et al., 2017). Defines test methodologies for data center network functions including incast, ECN marking, and lossless behavior. The AI fabric companion methodology documents extend RFC 8239 for distributed AI collective traffic patterns. |
| **RFC 2119 / RFC 8174** | "Key words for use in RFCs to Indicate Requirement Levels" (Bradner, 1997; Leiba, 2017). Define the normative requirement language: MUST, MUST NOT, REQUIRED, SHALL, SHALL NOT, SHOULD, SHOULD NOT, RECOMMENDED, MAY, and OPTIONAL. RFC 8174 clarifies that these terms are normative only when in uppercase; lowercase uses are not normative. |
{: #reference-standard title="Referenced Standards Abbreviations"}

# IANA Considerations

This document is a terminology document and has no IANA actions.

> The UET UDP destination port 4793 referenced in {{tab-rocev2}} is expected to be assigned through the Ultra Ethernet Specification {{UEC-SPEC-1.0}}; this document does not request any IANA assignment.

Reviewer: ```suggestion
Note that the UET UDP destination port 4793 referenced in {{tab-rocev2}} is expected to be assigned through the Ultra Ethernet Specification {{UEC-SPEC-1.0}}; this document does not request any IANA assignment.
```
It looks this paragaph provides additional clarification to the line 421, however I don't think this paragraph should be kept when this document is moving for publication since this document is IETF document instead of UEC document. maybe add this paragraph as Editor-note?

# Security Considerations

> This document defines benchmarking terminology and does not specify any protocol mechanism. It therefore introduces no new protocol-level security considerations beyond those of the underlying technologies it references. The following operational considerations apply to environments in which the terms defined herein are exercised:

Reviewer: Please refer to https://datatracker.ietf.org/doc/rfc8238/  and https://datatracker.ietf.org/doc/rfc8455/ to see how terminology draft write security consideration.

Author: thank you !  let me add a new PR for this (just the security section)

Author: please check commit edfb63cde82412c4c3d4e5aba5d4bcb2ef3d9171

> - Lab-only scope. As stated in Section 1.2, the methodology this terminology supports is intended for controlled laboratory environments. Direct application of these benchmarks to production fabrics carrying tenant traffic is out of scope and may expose operational systems to the failure modes described below.

Reviewer: If we use template defined in RFC8238, this paragraph is not needed.

Author: understood, please see the last commit with the new / updated security considerations ,
